File: Example/Objective-C/MZFormSheetPresentationController Objective-C Example/ViewController.m
//
//  ViewController.m
//  MZFormSheetPresentationViewController Objective-C Example
//
//  Created by Michal Zaborowski on 18.06.2015.
//  Copyright (c) 2015 Michal Zaborowski. All rights reserved.
//

#import "ViewController.h"
#import "MZFormSheetPresentationViewController.h"
#import "CustomTransition.h"
#import "PresentedTableViewController.h"
#import "MZFormSheetPresentationViewControllerSegue.h"
#import "CustomPrenstationControllerAnimator.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Set Default Background collor for all presentation controllers
    [[MZFormSheetPresentationController appearance] setBackgroundColor:[[UIColor darkGrayColor] colorWithAlphaComponent:0.3]];

    [MZTransition registerTransitionClass:[CustomTransition class] forTransitionStyle:MZFormSheetPresentationTransitionStyleCustom];
}

- (UINavigationController *)formSheetControllerWithNavigationController {
    return [self.storyboard instantiateViewControllerWithIdentifier:@"formSheetController"];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"segue"]) {
        MZFormSheetPresentationViewControllerSegue *presentationSegue = (id)segue;
        presentationSegue.formSheetPresentationController.presentationController.shouldApplyBackgroundBlurEffect = YES;
        UINavigationController *navigationController = (id)presentationSegue.formSheetPresentationController.contentViewController;
        PresentedTableViewController *presentedViewController = [navigationController.viewControllers firstObject];
        presentedViewController.textFieldBecomeFirstResponder = YES;
        presentedViewController.passingString = @"PASSSED DATA!!";
    }
}

#pragma mark - Mix

- (void)passDataToViewControllerAction {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.shouldDismissOnBackgroundViewTap = YES;
    formSheetController.presentationController.shouldApplyBackgroundBlurEffect = YES;
    
    PresentedTableViewController *presentedViewController = [navigationController.viewControllers firstObject];
    presentedViewController.textFieldBecomeFirstResponder = YES;
    presentedViewController.passingString = @"PASSSED DATA!!";
    
    formSheetController.willPresentContentViewControllerHandler = ^(UIViewController *vc) {
        UINavigationController *navigationController = (id)vc;
        PresentedTableViewController *presentedViewController = [navigationController.viewControllers firstObject];
        [presentedViewController.view layoutIfNeeded];
        presentedViewController.textField.text = @"PASS DATA DIRECTLY TO OUTLET!!";
    };
    
    [self presentViewController:formSheetController animated:YES completion:nil];
    
}
- (void)blurEffectAction {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.interactivePanGestureDismissalDirection = MZFormSheetPanGestureDismissDirectionAll;
    formSheetController.presentationController.shouldApplyBackgroundBlurEffect = YES;
    formSheetController.presentationController.blurEffectStyle = UIBlurEffectStyleLight;
    
    [self presentViewController:formSheetController animated:YES completion:nil];
}
- (void)parallaxEffectAction {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.shouldUseMotionEffect = YES;
    
    [self presentViewController:formSheetController animated:YES completion:nil];
}
- (void)customContentViewSizeAction {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.contentViewSize = CGSizeMake(100, 100);
    
    [self presentViewController:formSheetController animated:YES completion:nil];
}

- (void)customBackgroundColorAction {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.backgroundColor = [[UIColor redColor] colorWithAlphaComponent:0.5];
    
    [self presentViewController:formSheetController animated:YES completion:nil];
}

- (void)centerVerticallyAction {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.shouldCenterVertically = YES;
    PresentedTableViewController *presentedViewController = [navigationController.viewControllers firstObject];
    presentedViewController.textFieldBecomeFirstResponder = YES;
    
    __weak typeof(formSheetController) weakFormSheet = formSheetController;
    formSheetController.presentationController.frameConfigurationHandler = ^(UIView * __nonnull presentedView, CGRect currentFrame, BOOL isKeyboardVisible) {
        if (UIInterfaceOrientationIsLandscape([UIApplication sharedApplication].statusBarOrientation)) {
            return CGRectMake(CGRectGetMidX(weakFormSheet.presentationController.containerView.bounds) - 210, currentFrame.origin.y, 420, currentFrame.size.height);
        }
        
        return currentFrame;
    };
    
    [self presentViewController:formSheetController animated:YES completion:nil];
}

- (void)contentViewShadowAction {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.interactivePanGestureDismissalDirection = MZFormSheetPanGestureDismissDirectionUp | MZFormSheetPanGestureDismissDirectionDown | MZFormSheetPanGestureDismissDirectionLeft | MZFormSheetPanGestureDismissDirectionRight;

    __weak MZFormSheetPresentationViewController *weakController = formSheetController;
    formSheetController.willPresentContentViewControllerHandler = ^(UIViewController *a) {
        weakController.contentViewCornerRadius = 5.0;
        weakController.shadowRadius = 6.0;
    };
    
    [self presentViewController:formSheetController animated:YES completion:nil];
}

- (void)twoFormSheetControllersAction {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.shouldDismissOnBackgroundViewTap = YES;
    formSheetController.presentationController.portraitTopInset = 10;
    
    [self presentViewController:formSheetController animated:YES completion:^{
        UIViewController *controller = [self.storyboard instantiateViewControllerWithIdentifier:@"TransparentViewController"];
        MZFormSheetPresentationViewController *formSheetController2 = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:controller];
        formSheetController2.presentationController.shouldDismissOnBackgroundViewTap = YES;
        formSheetController2.presentationController.shouldApplyBackgroundBlurEffect = YES;
        [formSheetController presentViewController:formSheetController2 animated:YES completion:nil];
    }];
}

- (void)customPresentationControllerAnimator {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.animatorForPresentationController = [[CustomPrenstationControllerAnimator alloc] init];
    formSheetController.presentationController.backgroundColor = [UIColor clearColor];
    formSheetController.presentationController.shouldDismissOnBackgroundViewTap = YES;
    formSheetController.presentationController.transparentTouchEnabled = YES;

    [self presentViewController:formSheetController animated:YES completion:nil];
}

- (void)transparentBackgroundViewAction {
    UIViewController *viewController = [self.storyboard instantiateViewControllerWithIdentifier:@"TransparentViewController"];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:viewController];
    formSheetController.presentationController.transparentTouchEnabled = NO;
    // Uncomment if you don't want to capture status bar appearance in presentedViewController
//    formSheetController.modalPresentationCapturesStatusBarAppearance = NO;
    
    [self presentViewController:formSheetController animated:YES completion:nil];
}

- (void)panGestureDismissingGesture {
//    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    UIViewController *viewController = [self.storyboard instantiateViewControllerWithIdentifier:@"TransparentViewController"];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:viewController];
    formSheetController.interactivePanGestureDismissalDirection = MZFormSheetPanGestureDismissDirectionAll;
    formSheetController.allowDismissByPanningPresentedView = YES;

    [self presentViewController:formSheetController animated:YES completion:nil];
}

- (void)presentFormSheetControllerWithTransition:(NSInteger)transition {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.shouldDismissOnBackgroundViewTap = YES;
    formSheetController.contentViewControllerTransitionStyle = (MZFormSheetPresentationTransitionStyle)transition;
    
    [self presentViewController:formSheetController animated:(transition != MZFormSheetPresentationTransitionStyleNone) completion:nil];
}

- (void)presentFormSheetControllerWithKeyboardMovement:(NSInteger)movementOption {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.shouldDismissOnBackgroundViewTap = YES;
    formSheetController.presentationController.movementActionWhenKeyboardAppears = (MZFormSheetActionWhenKeyboardAppears)movementOption;
    formSheetController.contentViewControllerTransitionStyle = MZFormSheetPresentationTransitionStyleFade;
    PresentedTableViewController *presentedViewController = [navigationController.viewControllers firstObject];
    presentedViewController.textFieldBecomeFirstResponder = YES;
    
    // Always on center
    if (movementOption == MZFormSheetActionWhenKeyboardAppearsCenterVertically) {
        formSheetController.presentationController.shouldCenterVertically = YES;
    }
    
    [self presentViewController:formSheetController animated:YES completion:nil];
}

- (void)modalPresentationInsideFormSheetController {
    UINavigationController *navigationController = [self formSheetControllerWithNavigationController];
    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentViewController:navigationController];
    formSheetController.presentationController.shouldDismissOnBackgroundViewTap = YES;
    formSheetController.presentationController.portraitTopInset = 10;
    
    // This need to be set because presented modaly view need to be masked
    formSheetController.view.layer.masksToBounds = YES;
    
    [self presentViewController:formSheetController animated:YES completion:^{
        UIViewController *controller = [self.storyboard instantiateViewControllerWithIdentifier:@"TransparentViewController"];
        controller.modalPresentationStyle = UIModalPresentationOverCurrentContext;
        controller.view.layer.cornerRadius = navigationController.view.layer.cornerRadius;
        
        // Imporatant is to present view controller from navigationController
        // because UITransitionView will be added to navigationController view !!!
        [navigationController presentViewController:controller animated:YES completion:nil];
    }];
}

- (void)formSheetView {
    UIImageView *image = [[UIImageView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, 280.0f, 187.0f)];
    image.image = [UIImage imageNamed:@"home"];

    MZFormSheetPresentationViewController *formSheetController = [[MZFormSheetPresentationViewController alloc] initWithContentView:image];
    formSheetController.presentationController.shouldCenterVertically = YES;
    formSheetController.presentationController.shouldDismissOnBackgroundViewTap = YES;

    [self presentViewController:formSheetController animated:YES completion:nil];
}

#pragma mark - Table View

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    if (indexPath.section == 1) {
        switch (indexPath.row) {
            case 0: [self passDataToViewControllerAction]; break;
            case 1: [self blurEffectAction]; break;
            case 2: [self parallaxEffectAction]; break;
            case 3: [self customContentViewSizeAction]; break;
            case 4: [self customBackgroundColorAction]; break;
            case 5: /* Storyboard Segue */ break;
            case 6: [self centerVerticallyAction]; break;
            case 7: [self contentViewShadowAction]; break;
            case 8: [self twoFormSheetControllersAction]; break;
            case 9: [self transparentBackgroundViewAction]; break;
            case 10: [self customPresentationControllerAnimator]; break;
            case 11: [self panGestureDismissingGesture]; break;
            case 12: [self modalPresentationInsideFormSheetController]; break;
            case 13: [self formSheetView]; break;
            default:
                break;
        }
    }
    
    if (indexPath.section == 2) {
        [self presentFormSheetControllerWithTransition:indexPath.row];
    } else if (indexPath.section == 3) {
        [self presentFormSheetControllerWithKeyboardMovement:indexPath.row];
    }
}

@end

File: Example/Swift/MZFormSheetPresentationController Swift Example/ViewController.swift
//
//  ViewController.swift
//  MZFormSheetPresentationController Swift Example
//
//  Created by Michal Zaborowski on 18.06.2015.
//  Copyright (c) 2015 Michal Zaborowski. All rights reserved.
//

import UIKit

class ViewController: UITableViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        MZTransition.registerClass(CustomTransition.self, for: .custom)
    }

    func formSheetControllerWithNavigationController() -> UINavigationController {
        return self.storyboard!.instantiateViewController(withIdentifier: "formSheetController") as! UINavigationController
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let identifier = segue.identifier {
            if identifier == "segue" {
                let presentationSegue = segue as! MZFormSheetPresentationViewControllerSegue
                presentationSegue.formSheetPresentationController.presentationController?.shouldApplyBackgroundBlurEffect = true
                let navigationController = presentationSegue.formSheetPresentationController.contentViewController as! UINavigationController
                let presentedViewController = navigationController.viewControllers.first as! PresentedTableViewController
                presentedViewController.textFieldBecomeFirstResponder = true
                presentedViewController.passingString = "PASSED DATA"
            }
        }
    }
    
    // MARK: Mix
    
    func passDataToViewControllerAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.shouldDismissOnBackgroundViewTap = true
        formSheetController.presentationController?.shouldApplyBackgroundBlurEffect = true
        
        let presentedViewController = navigationController.viewControllers.first as! PresentedTableViewController
        presentedViewController.textFieldBecomeFirstResponder = true
        presentedViewController.passingString = "PASSED DATA"
        
        formSheetController.willPresentContentViewControllerHandler = { vc in
            let navigationController = vc as! UINavigationController
            let presentedViewController = navigationController.viewControllers.first as! PresentedTableViewController
            presentedViewController.view?.layoutIfNeeded()
            presentedViewController.textField?.text = "PASS DATA DIRECTLY TO OUTLET!!"
        }
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func blurEffectAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.shouldApplyBackgroundBlurEffect = true
        formSheetController.presentationController?.blurEffectStyle = UIBlurEffect.Style.dark
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func parallaxEffectAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.shouldUseMotionEffect = true
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func customContentViewSizeAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.contentViewSize = CGSize(width: 100, height: 100)
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func compressedContentViewSizeAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.contentViewSize = UIView.layoutFittingCompressedSize
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func expandedContentViewSizeAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.contentViewSize = UIView.layoutFittingExpandedSize
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func customBackgroundColorAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.backgroundColor = UIColor.red.withAlphaComponent(0.3)
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func centerVerticallyAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.shouldCenterVertically = true
        let presentedViewController = navigationController.viewControllers.first as! PresentedTableViewController
        presentedViewController.textFieldBecomeFirstResponder = true
        
        formSheetController.presentationController?.frameConfigurationHandler = { [weak formSheetController] view, currentFrame, isKeyboardVisible in
            if UIApplication.shared.statusBarOrientation.isLandscape {
                return CGRect(x: formSheetController!.presentationController!.containerView!.bounds.midX - 210, y: currentFrame.origin.y, width: 420, height: currentFrame.size.height)
            }
            return currentFrame
        };
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func contentViewShadowAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        
        formSheetController.willPresentContentViewControllerHandler = { [weak formSheetController] (value: UIViewController)  -> Void in
            if let weakController = formSheetController {
                weakController.contentViewCornerRadius = 5.0
                weakController.shadowRadius = 6.0
            }
        }
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func twoFormSheetControllersAction() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.shouldDismissOnBackgroundViewTap = true
        formSheetController.presentationController?.portraitTopInset = 10
        self.present(formSheetController, animated: true, completion: {
            let navigationController = self.formSheetControllerWithNavigationController()
            let formSheetController2 = MZFormSheetPresentationViewController(contentViewController: navigationController)
            formSheetController2.presentationController?.shouldDismissOnBackgroundViewTap = true
            formSheetController2.presentationController?.shouldApplyBackgroundBlurEffect = true
            formSheetController.present(formSheetController2, animated: true, completion: nil)
        })
    }
    
    func transparentBackgroundViewAction() {
        let viewController = self.storyboard!.instantiateViewController(withIdentifier: "TransparentViewController")
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: viewController)
        formSheetController.presentationController?.isTransparentTouchEnabled = false
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    func panGestureDismissingGesture() {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.interactivePanGestureDismissalDirection = .all;
        self.present(formSheetController, animated: true, completion: nil)
    }

    func formSheetView() {
        let label = UILabel(frame: CGRect(origin: CGPoint.zero, size: CGSize(width: 200.0, height: 25.0)))
        label.backgroundColor = .blue
        label.text = "Testing with just a view"
        label.textAlignment = .center
        label.textColor = .white

        let formSheetController = MZFormSheetPresentationViewController(contentView: label)
        if let presentationController = formSheetController.presentationController {
            presentationController.shouldCenterVertically = true
            presentationController.shouldDismissOnBackgroundViewTap = true
        }

        self.present(formSheetController, animated: true, completion: nil)
    }
    
    // MARK: -
    
    func presentFormSheetControllerWithTransition(transition: Int) {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.shouldDismissOnBackgroundViewTap = true
        formSheetController.contentViewControllerTransitionStyle = MZFormSheetPresentationTransitionStyle(rawValue: transition)!
        
        self.present(formSheetController, animated: MZFormSheetPresentationTransitionStyle(rawValue: transition)! != .none, completion: nil)
    }
    
    func presentFormSheetControllerWithKeyboardMovement(movementOption: Int) {
        let navigationController = self.formSheetControllerWithNavigationController()
        let formSheetController = MZFormSheetPresentationViewController(contentViewController: navigationController)
        formSheetController.presentationController?.shouldApplyBackgroundBlurEffect = true
        formSheetController.presentationController?.movementActionWhenKeyboardAppears = MZFormSheetActionWhenKeyboardAppears(rawValue: movementOption)!
        formSheetController.contentViewControllerTransitionStyle = MZFormSheetPresentationTransitionStyle.fade
        let presentedViewController = navigationController.viewControllers.first as! PresentedTableViewController
        presentedViewController.textFieldBecomeFirstResponder = true
        
        self.present(formSheetController, animated: true, completion: nil)
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        if indexPath.section == 0 {
            switch indexPath.row {
            case 0: passDataToViewControllerAction()
            case 1: blurEffectAction()
            case 2: parallaxEffectAction()
            case 3: customContentViewSizeAction()
            case 4: compressedContentViewSizeAction()
            case 5: expandedContentViewSizeAction()
            case 6: customBackgroundColorAction()
            case 7: /* Storyboard segue */ break;
            case 8: centerVerticallyAction()
            case 9: contentViewShadowAction()
            case 10: twoFormSheetControllersAction()
            case 11: transparentBackgroundViewAction()
            case 12: panGestureDismissingGesture()
            case 13: formSheetView()
            default: break;
            }
        } else if indexPath.section == 1 {
            presentFormSheetControllerWithTransition(transition: indexPath.row)
        } else if indexPath.section == 2 {
            presentFormSheetControllerWithKeyboardMovement(movementOption: indexPath.row)
        }
    }
}

